File: 16. Sorting_Techniques/10. Shell_Sort/C/code.cpp
#include <iostream>
#include <cmath>
using namespace std;
void shellSort(int data[], int n) {
    // int gap = (n / 2);  // with floor
    int gap = ceil(n / 2); // with ceil
    while (gap > 0) {
        bool no_exchange;
        do {
            no_exchange = true;
            for (int i = 0; i < n - gap; ++i) {
                if (data[i] > data[i + gap]) {
                    // Swap elements
                    int temp = data[i];
                    data[i] = data[i + gap];
                    data[i + gap] = temp;
                    no_exchange = false;
                }
            }
        } while (!no_exchange);
        gap /= 2;
    }
}

int main() {
    // int data[] = {12, 34, 54, 2, 3}; 
    int data[] = {23,7,9,5,2,29,15,19,31}; 
    int n = sizeof(data) / sizeof(data[0]);
    
    shellSort(data, n); 
    
    cout << "Sorted array: ";
    for (int i = 0; i < n; ++i) {
        cout << data[i] << " ";
    };
    
    return 0;
}

File: 16. Sorting_Techniques/10. Shell_Sort/java/code.java
import java.lang.Math;
class ShellSort {
    public void sort(int array[]) {
      int n = array.length;
      for (int gap = n / 2; gap > 0; Math.ceil(gap /= 2)) {
      // for (int gap = n / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < n; i++) {
        //   System.out.print(gap);
          int key = array[i];
          int j = i;
          while (j >= gap && array[j - gap] > key) {
            array[j] = array[j - gap];
            j -= gap;
          }
          array[j] = key;
        }
      }
    }
  }
  

public class code
{
	public static void main(String[] args) {
	    int A[] = {9,5,16,8,13,6,12,10,4,2,3};
	    ShellSort s = new ShellSort();
	    for(int i = 0 ; i < 11 ; i++){
	        System.out.print(A[i] + " ");
	    }
	    s.sort(A);
	    System.out.println();
	    for(int i = 0 ; i < 11 ; i++){
	        System.out.print(A[i] + " ");
	    }
	}
}
